packages/autorest.go/src/m4togocodemodel/types.ts

```diff
@@ -544,20 +537,35 @@ function adaptLiteralValue(constSchema: m4.ConstantSchema): go.LiteralValue {
     case m4.SchemaType.String:
     case m4.SchemaType.Duration:
     case m4.SchemaType.Uuid: {
-      const keyName = `literal-string-${constSchema.value.value}`;
-      let literalString = types.get(keyName);
-      if (literalString) {
-        return <go.LiteralValue>literalString;
-      }
-      literalString = new go.LiteralValue(new go.PrimitiveType('string'), constSchema.value.value);
-      types.set(keyName, literalString);
-      return literalString;
+      return adaptStringLiteral(constSchema.value.value);
     }
     default:
       throw new Error(`unsupported scheam type ${constSchema.valueType.type} for LiteralValue`);
   }
 }
 
+function adaptConstLiteral(constType: go.ConstantType, constValue: go.ConstantValue): go.LiteralValue {
+  const keyName = `literal-${constType.name}-${constValue.value}`;
+  let literalConst = types.get(keyName);
+  if (literalConst) {
+    return <go.LiteralValue>literalConst;
+  }
+  literalConst = new go.LiteralValue(constType, constValue);
+  types.set(keyName, literalConst);
+  return literalConst;
+}
+
+function adaptStringLiteral(literal: string): go.LiteralValue {
+  const keyName = `literal-string-${literal}`;
+  let literalString = types.get(keyName);
+  if (literalString) {
+    return <go.LiteralValue>literalString;
+  }
+  literalString = new go.LiteralValue(new go.PrimitiveType('string'), literal);
```

Comment:
Is it this need for a `PrimitiveType` that prevents writing a function that handles both string and const literals?

Reply:
That and the type cache key names are slightly different.

Reply:
I'm just looking at this thinking, if there were a type for string literals having `name` "string" and some way to produce a `PrimtiveType`, then `adaptStringLiteral` could collapse into `adaptConstLiteral`. Would that be painfully complex or semantically weird?

Reply:
I've refactored this so it's consolidated within `getDiscriminatorLiteral`. I reverted the other part of the changes which is adapting M4 to Go code model. Part of what was making things complicated earlier was trying to mix "adapting code" with "adapted code". We should keep those kinds of things separate.